Build and plugin hooks refactoring for code splitting and asset emission workflows

guybedford opened the pull request:
This is the rework of #2126.

Resolves #2163, #2092, #1611, #1629, #1282, #803.

### Refactoring Summary (for the third time...):

* transformBundle has been renamed to transformChunk taking one extra argument - the chunk itself
* I've added buildStart and buildEnd hooks, that apply for the `rollup.rollup` lifecycle, regardless of generate or write calls. `buildEnd` takes an error parameter if there was a build error.
* The plugin context is extended to include `this.resolveId` and `this.isExternal`, as well as the existing `this.warn` and `this.error` and now applies equally for all plugin hooks
* `entryNames` and `chunkNames` are now `entryFileNames` and `chunkFileNames`.
* Code splitting is refactored into a single path based on a `inlineDynamicImports` input option, which by default is set to try for a single string input and false otherwise

### Asset Emission Features

The goal here is to support workflows of building assets like CSS alongside chunks.

Use cases such as:

* Being able to work out only which CSS is associated with a given chunk
* Being able to get the URL of an asset in a straightfoward way from within module code
* Being able to tell which exports have been tree-shaken from a module to prune CSS (eg treeshaking for CSS modules)
* Being able to run in-memory builds of chunks and assets and get both of their sources without having to write the files - useful for server side rendering style workflows

I've really gone back and fourth on a lot of these quite a bit to try and have Rollup do the minimal work necessary to keep maintenance down and to just open the right doors.

This PR provides:

* A new `assetFileNames` option is added, which by default is set to `assets/[name]-[hash][ext]`.
* Plugin context provides `emitAsset(name: string, source?: string | Buffer) => assetId: string`, `setAssetSource(assetId: string, source: string | Buffer)` and `getAssetFileName(assetId: string)`. This way an asset can be defined by a plugin at any time, and assets can also have their sources deferred until after the chunking and treeshaking so that they can perform optimizations based on matching the chunking and treeshaking behaviours. `getAssetFileName` can only be called after both the asset source is set and generate has been called, throwing an error otherwise.
* A special token `import.meta.ROLLUP_ASSET_URL_[assetid]` is supported in code which is replaced in the build just like `import.meta.url` but instead pointing to the URL of the given asset. Getting the AssetID early here is important to be able to emit these tokens in say `translate`, while the asset source itself might only be populated afterwards.
* A `generateBundle(outputOptions, outputBundle, isWrite)` hook is provided ideally to replace `onwrite` and `ongenerate` in the deprecation path. The `outputBundle` takes the form of `{ [fileName: string]: Chunk | string | Buffer }`, where a string or buffer indicates an asset.
* The Chunk.modules list is updated to no longer be an array of module Ids, but an object of the form `{ [moduleId: string]: { renderedExports: string[], removedExports: string[], renderedLength: number, originalLength: number } }`. This way it is possible to determine what exports have been tree-shaken, as well as workout out how much source reduction that corresponds to.
* `rollup.generate` and `rollup.write` still return the direct chunk in the single-file build case (being deprecated), and return the outputBundle object now by default when experimentalCodeSplitting is enabled.

I think the above finally gets us these use cases in a relatively simple and straightforward way for Rollup, allowing plugins to explore the space further and drawing a clear line on some of these hard problems so that we don't fall too deep into the rabbit hole on these things.

@lukastaegert ready for review finally :)

lukastaegert commented on src/rollup/types.d.ts:
```diff
@@ -265,10 +322,10 @@ export interface RollupCache {
 }
 
-export interface Bundle {
+export interface RollupFileBuild {
```
This name confused me at first as we are always building files. Maybe `RollupSingleFileBuild` as the "single" part ?

lukastaegert commented on src/ast/nodes/MemberExpression.ts:
```diff
@@ -222,14 +222,6 @@ export default class MemberExpression extends NodeBase {
-				this.property instanceof Identifier && this.property.name,
-				options.format
-			);
-			if (importMetaMechanism) code.overwrite(this.start, this.end, importMetaMechanism);
```
I think we already touched this in one of the previous reviews but great this could be moved to `MetaProperty`!

lukastaegert commented on src/finalisers/amd.ts:
```diff
@@ -16,7 +16,7 @@ export default function amd(
 		outro,
 		dynamicImport,
-		importMeta,
+		hasImportMetaUrl,
```
👍

lukastaegert commented on src/utils/transform.ts:
```diff
@@ -1,20 +1,94 @@
+				code: 'PLUGIN_WARNING',
+				id,
+				source,
+				pluginName: plugin.name
```
If a plugin does not have a name (which happens often for ad-hoc plugins) the warning will read: "(!) undefined plugin:  ..." which is somewhat confusing. One approach could be to set the `pluginName` to "unnamed" which makes for less confusing warnings.

The same would apply to errors though in that case, the error message is less confusing (the message seems to assume this is not a plugin error). Still, I think if the error reads "[!] (unnamed plugin) Error: ...", this would provide the helpful information that the error did not originate from rollup itself but probably from some code the user wrote themselves.

lukastaegert replied:
Another issue I stumbled upon: When I generate warnings or errors from the `transformBundle` hook, I get a maximum callstack error instead.

lukastaegert commented on src/utils/transformChunk.ts:
```diff
@@ -0,0 +1,67 @@
+		return promise.then(code => {
+			return Promise.resolve()
+				.then(() =>
+					(plugin.transformChunk || plugin.transformBundle).call(
```
I must admit I could not get `transformChunk` to be called at all in my experiments. A test to take some inspiration from would be really helpful.

lukastaegert commented on src/utils/transformChunk.ts:
```diff
@@ -0,0 +1,67 @@
+	options: OutputOptions
+) {
+	return graph.plugins.reduce((promise, plugin) => {
+		if (!plugin.transformBundle) return promise;
```
Actually, this might be the reason why `transformChunk` does not work at all...

guybedford replied:
Ahh, thanks.

lukastaegert commented on src/rollup/index.ts:
```diff
@@ -124,357 +102,375 @@ function getInputOptions(rawInputOptions: GenericConfigObject): any {
 
-		timeStart('BUILD', 1);
+		for (let plugin of graph.plugins) {
+			if (plugin.buildStart) plugin.buildStart.call(graph.pluginContext, inputOptions);
```
Using `this.warn` in the `buildStart` and `buildEnd` hooks also caused maximum callstack errors. I think we should have a few more simple tests for the plugin context API. There are already some tests in the `function` category for `this.warn` that could probably be easily adapted for the other hooks.

guybedford commented:
I'll integrate the feedback and some further changes I'd like to make on this tomorrow, so no need to review further before then.

guybedford commented on src/Graph.ts:
```diff
@@ -107,18 +119,78 @@ export default class Graph {
+					});
+				asset.source = source;
+			},
+			resolveId: undefined,
```
Sure will add one. There is a circular risk here too in calling this from the resolve hook itself, but I think that's just common sense to avoid.

guybedford commented on src/Graph.ts:
```diff
@@ -107,18 +119,78 @@ export default class Graph {
+		};
+
+		this.pluginContext = {
+			isExternal: this.isExternal,
```
👍

guybedford commented on src/Graph.ts:
```diff
@@ -107,18 +119,78 @@ export default class Graph {
+
+		this.pluginContext = {
+			isExternal: this.isExternal,
+			emitAsset: (name: string, source?: string | Buffer) => {
```
The reason it's a two-step process is due to the order of information:

1. You may want to construct an asset (say a CSS file) to match the list of modules defined by a chunk, so you will only know the asset source after generate is called.
2. The chunk source itself might want to reference the asset URL via `import.meta.ROLLUP_ASSET_URL_sd9f8fg0s`, even though the asset source isn't yet known at this point.
3. Similarly, assets themselves might need to reference other assets, and only be generated late.

So the separation of assetId through the three methods allows us to dodge all these issues!

lukastaegert replied:
Ah I see. import.meta.ROLLUP_ASSET_URL was one of the few things I had not tested yet. That makes sense!

guybedford commented:
Thanks for testing out those cases - I've fixed and included tests for these now.

I also added support for even later asset emissions at the generateBundle phase, along with a `getAssetFileName` function.

Ready for further review!

lukastaegert commented on src/rollup/index.ts:
```diff
@@ -124,357 +103,389 @@ function getInputOptions(rawInputOptions: GenericConfigObject): any {
+		timeStart('BUILD', 1);
 
-		if (!codeSplitting) {
+		if (!inputOptions.experimentalCodeSplitting) {
```
Very nice and clean to have all these validations grouped together. I wonder if it would be possible to move these option validations out of the rollup function and into `getInputOptions`? @

lukastaegert commented on src/rollup/index.ts:
```diff
@@ -124,357 +103,389 @@ function getInputOptions(rawInputOptions: GenericConfigObject): any {
 
-					const generated: { [chunkName: string]: OutputChunk } = {};
+					// populate asset files into output
+					const assetFileNames = outputOptions.assetFileNames || 'assets/[name]-[hash][ext]';
```
It would be really nice if "[ext]"  would not contain the "." as in this case, I could use a pattern such as "[ext]/[name].[ext]" i.e. grouping different types of assets into different folders.

guybedford replied:
The issue with this is that in theory we support assets without an extension, so the `[ext]` pattern can be the empty string. Perhaps we could introduce a new variable - `[extname]` or something like that?

guybedford replied:
Note we can always move to supporting a function for these patterns as well, with a `getVar` function argument itself - 

```js
assetFileNames: (vars) => {
  if (vars.ext === '.css')
    return `assets/style/${vars.name}`;
  return `assets/${vars.name}`;
}
```

guybedford replied:
Note that this is the same output as the NodeJS `path.extname`, by inluding the dot.

Perhaps instead of `[ext]` it should be called `[extname]` with `[ext]` being the version without the dot? That might make it more obvious to Node users too.

lukastaegert replied:
Sounds good

lukastaegert commented on test/hooks/index.js:
```diff
@@ -121,4 +279,325 @@ describe('hooks', () => {
+			});
+	});
+
+	it('supports CommonJS asset urls', () => {
```
I tried this for `amd` and got something like

```javascript
define(function () { 'use strict';

	console.log(new URL((typeof process !== 'undefined' && process.versions && process.versions.node ? 'file:' : '') + module.uri + '/../assets/asset.ext').href);

});
```

I would have expected `module` to be imported by the amd loader similar to how it is done for `import.meta.url`.

Also, `system` is not supported at all as a format but throws in a rather non-pleasing way with a TypeError.

guybedford replied:
Thanks, will fix.

guybedford commented on test/hooks/index.js:
```diff
@@ -121,4 +279,325 @@ describe('hooks', () => {
+			.rollup({
+				input: 'input',
+				experimentalCodeSplitting: true,
+				experimentalDynamicImport: true,
```
I think we can unflag `experimentalDynamicImport`, but I'm still weary of the dynamic import hook being documented.

guybedford commented on test/hooks/index.js:
```diff
@@ -5,6 +5,164 @@ const { loader } = require('../utils.js');
+			});
+	});
+
+	it('supports warnings in buildStart and buildEnd hooks', () => {
```
Thanks

guybedford commented:
I've updated these now. In adding tests for the System output I caught a magic string bug, which this now depends on at https://github.com/Rich-Harris/magic-string/pull/144.

guybedford commented:
I've tweaks the hooks slightly with the following additional changes (updated in the PR summary as well):
* buildStart and buildEnd now support promises
* processChunks is removed
* generateBundle now gets a chunk with a "modules" property that is no longer an array of moduleIds but instead an object of `{ [moduleId: string]: { renderedExports: string[], removedExports: string[], originalLength: number, renderedLength: number } }` allowing inspection into the treeshaking improvements.

Note that the `renderedLength` doesn't include the Rollup wrapper or imports or exports as well, so will always be less and may be slightly deceving (eg `export var p = 5;` is reported as a length of 10. We could potentially have a wrapperLength property on the chunk itself to correct this.

I also created https://github.com/Rich-Harris/magic-string/pull/145 for this to try optimize getting the length a bit.

lukastaegert commented:
> generateBundle now gets a chunk with a "modules" property that is no longer an array of moduleIds but instead an object of { [moduleId: string]: { renderedExports: string[], removedExports: string[], originalLength: number, renderedLength: number } } allowing inspection into the treeshaking improvements

I love this very much. This could be the first step to enable a whole now class of analysis and visualization plugins.

guybedford commented on src/ast/nodes/ExportDefaultDeclaration.ts:
```diff
@@ -159,7 +159,7 @@ export default class ExportDefaultDeclaration extends NodeBase {
 		);
 		const hasTrailingSemicolon = code.original.charCodeAt(this.end - 1) === 59; /*";"*/
 		if (systemBinding) {
-			code.appendLeft(
```
It's particularly the system binding case that suffers from the double write issue, where we want to add an appendage first, then later on rewrite the inner content. It might be more obvious think in terms of `overwriteInner` and `overwriteOuter` here, see https://github.com/Rich-Harris/magic-string/pull/144#issuecomment-393532572 for further comment on this.

guybedford commented on src/rollup/index.ts:
```diff
@@ -124,357 +108,321 @@ function getInputOptions(rawInputOptions: GenericConfigObject): any {
-							message:
-								'UMD and IIFE output formats are not supported with the experimentalCodeSplitting option.'
-						});
+					if (inputOptions.experimentalCodeSplitting) {
```
Exactly, I've turned it into a flag in this PR that we can then just disable when we are ready.

In terms of a checklist for launch, I think it would just be nice to have a little bit more feedback on usage patterns, to give us a bit more space to not worry about any unexpected API churn for a little while.

But I do think this gets us to almost stable... think of it as a release candidate.

guybedford replied:
The other thing to note is that adding the flag is a breaking change in that imports, exports and modules as properties of the build bundle before generate are no longer present.

lukastaegert replied:
Ah I see, yes that would be a breaking change. So maybe we keep it at least for the 0.60.0 and then remove it for the 1.0.0. One thing I keep wondering about is if it is a wise idea to directly return a map of the chunks from generate. This means we will never be able to add any other meta-information that is not chunk-specific here. Maybe instead we could return an object
`{chunks: {[chunk]: <chunk-data>}}` here? That would leave the door open for extensions.

lukastaegert replied:
One thing we could do for the 0.60.0, however, is to change *all* our tests to use `experimentalCodeSplitting` and thus test the new code paths. Should be a great way to find lingering bugs and inconsistencies. I fear I will not be able to finish the review this week as I had planned as I will be on a hiking trip until Sunday but it looks really great!

guybedford commented on src/rollup/index.ts:
```diff
@@ -124,357 +108,321 @@ function getInputOptions(rawInputOptions: GenericConfigObject): any {
+					) {
+						singleInputChunk = chunks.find(chunk => chunk.entryModule !== undefined);
+						imports = singleInputChunk.getImportIds();
+						exports = singleInputChunk.getExportNames();
```
At one point I was sharing this to avoid duplicate work, but now that it isn't that makes sense, thanks.

lukastaegert commented on test/function/samples/import-meta-url-compact/_config.js:
```diff
@@ -0,0 +1,20 @@
+const assert = require('assert');
+const path = require('path');
+const URL = require('url-parse');
```
url-parse needs to be added to rollup's dev dependencies. This is why tests on CI have been failing for some time.

guybedford commented:
I've adjusted the return value of generate and write to be `{ output: { [fileName: string]: string | Buffer | Chunk } }`, so it can be extensible in future.

Good idea on switching the tests to `experimentalCodeSplitting` for 0.60. Do you want to aim to include that in this PR as well?

Added url-parse as a dependency now, hopefully that does fix the tests, thanks was wondering what was up with that.

lukastaegert commented:
While playing around with the dynamic import inlining, I discovered the following bug (which actually seems to have existed for some time):

```javascript
// foo.js
export const foo = 1;

// main.js
import { foo } from './foo.js';
export const bar = 2
import('./foo.js')

// output when bundling main.js
const foo = 1;

var foo$1 = /*#__PURE__*/Object.freeze({
	foo: foo
});

const bar = 2;
Promise.resolve().then(function () { return foo$1; });

export { foo };
```

Note that the bundle is actually exporting `foo`, not `bar` i.e. the exports of `main.js` seem to have been replaced by the exports of `foo.js`. If you remove the non-dynamic import of `foo.js`, `bar` is correctly exported. Did not dig deeper into this yet.

lukastaegert commented:
> I've adjusted the return value of generate and write
👍

> Good idea on switching the tests to experimentalCodeSplitting for 0.60. Do you want to aim to include that in this PR as well?

Considering that I would love to have this PR in 0.60.0 and that there are now some important fixes in master that should be released soon, I would actually postpone this. Once this is released, we could just do the actual switch to remove the flag and would not have to edit all config files twice.

If you could have a look at the bug I outlined above, that would be really cool, though. Seems to be more like your area of expertise but I could have a look later myself as well.

guybedford commented:
Thanks for spotting the `inlineDynamicImports` bug, I've included the fix here in the latest commit - it was misdetecting the entry point module of the chunk, much more straightforward than it seemed :)
